packages/react-i18n/src/index.tsx

```diff
@@ -2,39 +2,35 @@
  * External dependencies
  */
 import * as React from 'react';
-import { __, _n, _nx, _x } from '@wordpress/i18n';
+import memize from 'memize';
+import { createI18n, I18n, LocaleData } from '@wordpress/i18n';
 import { createHigherOrderComponent } from '@wordpress/compose';
 
-const DEFAULT_LOCALE = 'en';
-
-export interface I18nProps {
-	__: typeof __;
-	_n: typeof _n;
-	_nx: typeof _nx;
-	_x: typeof _x;
+export interface I18nReact {
+	__: I18n[ '__' ];
+	_n: I18n[ '_n' ];
+	_nx: I18n[ '_nx' ];
+	_x: I18n[ '_x' ];
+	isRTL: I18n[ 'isRTL' ];
 	i18nLocale: string;
 }
 
-const I18nContext = React.createContext< I18nProps >( makeI18n( DEFAULT_LOCALE ) );
+const I18nContext = React.createContext< I18nReact >( makeContextValue() );
 
 interface Props {
-	/**
-	 * The current locale
-	 */
-	locale: string;
+	localeData?: LocaleData;
 }
-export const I18nProvider: React.FunctionComponent< Props > = ( { children, locale } ) => {
-	const [ contextValue, setContextValue ] = React.useState< I18nProps >( makeI18n( locale ) );
-	React.useEffect( () => {
-		setContextValue( makeI18n( locale ) );
-	}, [ locale ] );
-	return <I18nContext.Provider value={ contextValue }>{ children }</I18nContext.Provider>;
+export const I18nProvider: React.FunctionComponent< Props > = ( { children, localeData } ) => {
+	const makeStableContext = React.useRef< typeof makeContextValue >(
+		memize( makeContextValue, { maxSize: 1 } )
```

Comment:
Can `useMemo` possibly do the same job?
```js
const contextValue = useMemo( () => makeContextValue( localeData ), [ localeData ] );
```
Here, `contextValue` will be recomputed only if the `localeData` prop is not `===` to the previous value, and the memoization is per-component.

Reply:
> `makeStableContext` is a memoized version of `makeContextValue`, where
> 
> * the cache is not shared by component instances, but each instance has its own one
> * the cache has size one and remembers one latest return value?

Yes, this was my intention and I believe it's correct.

> Can `useMemo` possibly do the same job?

`useMemo` is [explicitly a performance optimization and makes no guarantees about behavior](https://reactjs.org/docs/hooks-reference.html#usememo).

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

That was my reasoning for avoiding `useMemo`. In reality, I think this qualifies as an optimization - I wouldn't expect any bugs if the react-i18n triggers a few extra renders.

Do you think we should go with `useMemo`?

Do you think we should switch to `useMemo`?

Reply:
> Do you think we should go with `useMemo`?

Yes, I think we have a perfect use case for `useMemo`. Our component also doesn't break if the context is recomputed. It just causes extra rerenders. But all components that use `useMemo` will have to rerender once `useMemo` decides to start forgetting things.